reflect: prevent AccidentallyQuadratic behaviour when decoding json s…

dgryski opened the pull request:
…tructs

dgryski committed:
reflect: prevent AccidentallyQuadratic behaviour when decoding json structs

dgryski commented:
I have a better hash function that drastically reduces the number of collisions on my test data.

dgryski commented on src/reflect/type.go:
```diff
@@ -631,6 +632,66 @@ func rawStructFieldFromPointer(descriptor *structType, fieldType *rawType, data
+		// Cache miss; populate cache for this type.
+		// Clear out the old map so we have space to allocate a new one
+		f[idx].offsets = nil
+		offsets := make(map[int]uintptr)
```
Could probably preallocate this map so we don't hit the growth case.

dgryski commented on src/reflect/type.go:
```diff
@@ -631,6 +632,66 @@ func rawStructFieldFromPointer(descriptor *structType, fieldType *rawType, data
+var fieldOffsetCache fieldOffsetCacheType
+
+func (f *fieldOffsetCacheType) lookup(t *structType, n int) uintptr {
+	idx := uintptr(unsafe.Pointer(t))
```
Should probably extract this hash logic to its own function.

dgryski commented on src/reflect/type.go:
```diff
@@ -631,6 +632,66 @@ func rawStructFieldFromPointer(descriptor *structType, fieldType *rawType, data
+func xorshiftMult64(x uint64) uint64
+
+func hashptr(p uintptr) uintptr {
+	if unsafe.Sizeof(p) == 32 {
```
Oops, this should be `4`.

dgryski commented:
PTAL.  Fixed all the issues I identified.

dgryski commented:
Huh. CI failed.

dgryski commented:
Two hard problems in computer science. 

When filling a cache, we need to call field.Size(), but for some structs that field was itself another struct which collided with the struct we were in the middle of trying to process, which would overwrite the cache entry and make things Bad.  Ouch.

dgryski committed:
reflect: ensure no recursive field offset cache filling collisions

Sometimes during a cache fill operation, a call to field.Size() can generate a
new call to rawField() which can potentially collide with the entry being
filled in.  This can corrupt the cache.  Add a simple bool lock to prevent
this.

dgryski committed:
reflect: use field offset cache instead of calculating by hand

deadprogram commented:
Seems like real CI error https://github.com/tinygo-org/tinygo/actions/runs/4530584804/jobs/7980277801?pr=3598#step:7:39

dgryski commented:
Yup and I know why it happened. This is a collision when the cache is being re populated but it's for the type that was in that slot. Since clearing the offsets map is the first thing we do, the other collision reads from a null map which is zero. Need to pull the locked check out of the big if completely.

dgryski commented:
Failing windows stdlib test is just the `net` package this time.

dgryski commented:
All green. PTAL.

deadprogram commented:
I re-ran a couples times CI is green.

dgryski committed:
reflect: cache struct alignment calculations

This brings the time for decoding 1000 GitHub Repository
responses from 17s down to 5s, and a real-world improvement
of a single request from 150ms to 40ms.

dgryski commented on src/reflect/type.go:
```diff
@@ -631,6 +632,93 @@ func rawStructFieldFromPointer(descriptor *structType, fieldType *rawType, data
 	}
 }
 
+const fieldOffsetCacheSize = 16
```
We could probably scale this like
```
const fieldOffsetCacheSze = 4 * unsafe.Sizeof(uintptr(0))
```
so when building natively or on larger machines we get a larger default cache.  I was just trying to be really conservative here.

dgryski commented on src/reflect/type.go:
```diff
@@ -850,6 +918,8 @@ func (t *rawType) Size() uintptr {
 	}
 }
 
+var structAlignCache = make(map[*rawType]int)
```
We could probably added additional caches for `Size()`, `Comparable()`, and `isBinary()`. (assuming we didn't want to add Comparable and isBinary to a flag byte somewhere already)

kenbell commented on src/reflect/type.go:
```diff
@@ -631,6 +632,93 @@ func rawStructFieldFromPointer(descriptor *structType, fieldType *rawType, data
+	idx := hashStructTypePtr(t)
+	idx %= fieldOffsetCacheSize
+
+	if f[idx].locked {
```
I'm not sure about this locking primitive.  I don't think there's anything that guarantees a write to a bool is atomic like this.  There's no memory barrier or anything...

dgryski replied:
There isn't.  However, TinyGo also doesn't have any *actual* concurrency so we'd only be interrupted by, well, an interrupt.  This is mostly to solve the case of recursive calls from the same thread -- maybe "lock" was a bad name and "in-progress" or "incomplete" would be a better one.  I can replace this with something from the `sync` package if you'd like.

kenbell replied:
The recursion path isn't obvious (at least to me), perhaps just add some comments explaining why the lock exists and where the recursion can happen?

dgryski replied:
Updated the comment around `locked`.  PTAL

aykevl commented:
@dgryski have you run some size comparison tests? For example, `make smoke-test` in the drivers repo.

(I haven't really looked at the code yet).

dgryski commented:
The extra code complexity here is really to avoid having the cache grow without bound.  If we don't care about that we can probably just replace it with a straight map and move on.  That shouldn't impact the code size as much but will probably affect runtime memory when reflection is used because there are no longer any cache evictions.

dgryski commented:
Non-zero entries from `make smoke-test` and `sizediff`
```
153528 153968    440   0.3%  tinygo build -size short -o ./build/test.hex -target=feather-m4 ./examples/sdcard/tinyfs/
 95572  96020    448   0.5%  tinygo build -size short -o ./build/test.hex -target=wioterminal ./examples/rtl8720dn/mqttsub/
132540 133020    480   0.4%  tinygo build -size short -o ./build/test.hex -target=arduino-nano33 ./examples/wifinina/ntpclient/main.go
132536 133016    480   0.4%  tinygo build -size short -o ./build/test.hex -target=arduino-nano33 ./examples/wifinina/udpstation/main.go
133056 133536    480   0.4%  tinygo build -size short -o ./build/test.hex -target=arduino-nano33 ./examples/wifinina/webclient/main.go
132748 133236    488   0.4%  tinygo build -size short -o ./build/test.hex -target=arduino-nano33 ./examples/wifinina/tcpclient/main.go
 69328  69824    496   0.7%  tinygo build -size short -o ./build/test.hex -target=wioterminal ./examples/rtl8720dn/webserver/
 79964  80468    504   0.6%  tinygo build -size short -o ./build/test.hex -target=wioterminal ./examples/rtl8720dn/webclient/
 20004  20548    544   2.7%  tinygo build -size short -o ./build/test.hex -target=microbit-v2 ./examples/microbitmatrix/main.go
 71140  71708    568   0.8%  tinygo build -size short -o ./build/test.hex -target=p1am-100 ./examples/p1am/main.go
 23984  24552    568   2.4%  tinygo build -size short -o ./build/test.uf2 -target=pico ./examples/ssd1289/main.go
 19956  20532    576   2.9%  tinygo build -size short -o ./build/test.hex -target=microbit ./examples/microbitmatrix/main.go
 66012  67060   1048   1.6%  tinygo build -size short -o ./build/test.hex -target=itsybitsy-m0 ./examples/flash/console/spi
 60836  61908   1072   1.8%  tinygo build -size short -o ./build/test.hex -target=itsybitsy-m0 ./examples/bmp280/main.go
 66452  67524   1072   1.6%  tinygo build -size short -o ./build/test.hex -target=itsybitsy-m0 ./examples/ds3231/main.go
 62316  63388   1072   1.7%  tinygo build -size short -o ./build/test.hex -target=pyportal ./examples/flash/console/qspi
 74812  75884   1072   1.4%  tinygo build -size short -o ./build/test.hex -target=feather-m0 ./examples/dht/main.go
 69424  70496   1072   1.5%  tinygo build -size short -o ./build/test.hex -target=feather-m4 ./examples/sdcard/console/
 58428  59508   1080   1.8%  tinygo build -size short -o ./build/test.hex -target=itsybitsy-m0 ./examples/adt7410/main.go
 60796  61876   1080   1.8%  tinygo build -size short -o ./build/test.hex -target=itsybitsy-m0 ./examples/bmi160/main.go
 64452  65532   1080   1.7%  tinygo build -size short -o ./build/test.hex -target=itsybitsy-m0 ./examples/mcp2515/main.go
 53952  55032   1080   2.0%  tinygo build -size short -o ./build/test.hex -target=microbit ./examples/sht3x/main.go
 53932  55012   1080   2.0%  tinygo build -size short -o ./build/test.hex -target=microbit ./examples/shtc3/main.go
 68316  69396   1080   1.6%  tinygo build -size short -o ./build/test.hex -target=xiao ./examples/pcf8563/alarm/
 67976  69056   1080   1.6%  tinygo build -size short -o ./build/test.hex -target=xiao ./examples/pcf8563/time/
 68252  69332   1080   1.6%  tinygo build -size short -o ./build/test.hex -target=xiao ./examples/pcf8563/timer/
 57660  58828   1168   2.0%  tinygo build -size short -o ./build/test.hex -target=feather-m4 ./examples/i2csoft/adt7410/
 78504  79672   1168   1.5%  tinygo build -size short -o ./build/test.hex -target=nucleo-wl55jc ./examples/lora/lorawan/atcmd/
263736 264912   1176   0.4%  tinygo build -size short -o ./build/test.hex -target=pyportal ./examples/ili9341/slideshow
 57892  59076   1184   2.0%  tinygo build -size short -o ./build/test.hex -target=feather-nrf52840 ./examples/is31fl3731/main.go
 64692  66132   1440   2.2%  tinygo build -size short -o ./build/test.hex -target=feather-m0 ./examples/gps/i2c/main.go
 22232  23676   1444   6.5%  tinygo build -size short -o ./build/test.hex -target=arduino-nano33 ./examples/espat/espconsole/main.go
 22432  23876   1444   6.4%  tinygo build -size short -o ./build/test.hex -target=arduino-nano33 ./examples/espat/esphub/main.go
 22240  23684   1444   6.5%  tinygo build -size short -o ./build/test.hex -target=arduino-nano33 ./examples/espat/espstation/main.go
 65096  66544   1448   2.2%  tinygo build -size short -o ./build/test.hex -target=feather-m0 ./examples/gps/uart/main.go
 19040  20508   1468   7.7%  tinygo build -size short -o ./build/test.hex -target=bluepill ./examples/ds1307/time/main.go
sum: 34504 (1.0%) geomean: 1.1%
```

Edit: technically this geomean calculation was incorrect because of all the 0s.  It should be closer effectively 0 also.

dgryski committed:
reflect: add structSizeCache and structIsBinaryCache

dgryski committed:
reflect: scale size of fieldOffsetCache with size of machine

deadprogram commented:
What is the status with this PR now?

dgryski committed:
reflect: expand comment for fieldOffsetCache lock

dgryski commented:
@aykevl Are you still concerned about the code size impact of this change?

aykevl reviewed (commented):
Most of these changes seem fine by me (even though I don't really like using hashmaps in the reflect package - it does work).

What I don't quite follow is `fieldOffsetCacheType`. Can you explain what it does, and why it is necessary?

One possible consequence that may be unexpected is that comparing two interfaces can now allocate, while previously I don't think they would allocate. This can have consequences for interrupt code for example.

dgryski commented:
I created `fieldOffsetCacheType` so I could put a method on it.  I suppose I could just declare the `fieldOffsetCache` as a variable of the appropriate type and call `lookup()` passing along the cache.  I don't know if there's a global "in interrupt" flag, but we could detect that and use the slow non-allocating path in that case.

Note that all this extra work is because we don't store struct sizes or field offsets in the binary, so we have to calculate them.  But all these loops end up being quite expensive in the case where user code is iterating through all the fields in a struct.  In exchange for a larger binary, we could replace all these caches and maps with direct lookups.

dgryski commented:
Should we prototype adding size and and offsets to structs and seeing the effect that has?

dgryski commented:
Planning on closing this in favour of the PR that just adds the sizes to the type structures. However the caches for IsBinary and Comparable seem useful.

aykevl commented:
I believe there are two bits left over in the meta byte at the start of the type struct. Perhaps you can put `isBinary` (and maybe `Comparable`) in there?

dgryski commented:
I didn't want to take bits in the meta byte without discussion, but yes that would be a good place for them. I will add that to the other struct-size PR.

dgryski commented:
Filed https://github.com/tinygo-org/tinygo/issues/3683 instead.
